### PR TITLE
[Blog] Simplifying releases with Rust

### PR DESCRIPTION
**Pull request by locks**
This is my very belated article for the blog post rotation. It describes at a very high level my journey in implementing `tool-new-release`, a command-line tool to help the ember core learning team automate their release process.

### CODE REVIEW

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+topic: ember
+bio: 'Senior Frontend Engineer, Ember Framework and Learning Core teams member'
+description:
+  'Ricardo Mendes discusses a command like tool built in Rust that automates the releases process for the Ember Core Learning Team'
+og:
+  image: /assets/images/posts/2021-07-13-effective-infrastructure-for-efficient-development-workflows/og-image.png
+---
+
+It's not trivial to release a new version of Ember. This process involves several different core
+teams, Framework, Data, CLI and Learning and several libraries and resources. In this post Ricardo will show you how they're building a tool in Rust to automate the process for the Core Learning Team.
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I'm trying to rephrase this somehow to not have so many consecutive `and` `and` `and` `and` but not sure exactly how.

```suggestion
It's not trivial to release a new version of Ember. This process involves several different core
teams: _Framework team_, _Data team_, _CLI team_ and _Learning team_ and, at the same time, quite a few libraries and resources. In this post Ricardo will show you how they're building a tool in Rust to automate the process for the Core Learning Team.
```

Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I'm not sure which person is used when writing in simplabs' blog: is it _impersonal_, _first person_ or _third person_?

```suggestion
It's not trivial to release a new version of Ember. This process involves several different core
teams, Framework, Data, CLI and Learning and several libraries and resources. In this post I/we/Ricardo/it will be shown will show you how we're/they're/it is being built building a tool in Rust to automate the process for the Core Learning Team.
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+<!--break-->
+
+![Effective infrastructure for efficient development workflows illustration](/assets/images/posts/2021-07-13-effective-infrastructure-for-efficient-development-workflows/illustration.svg#full)
+
+## How do Ember project releases work
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I don't if it's just me but I see somebody could get confused between Ember Project releasing work (like freeing work) and Ember Project doing software releases, but you can ignore this comment if it's only me interpreting this xD

```suggestion
## How do releases in _Ember project_ work
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+An Ember project release involves 4 different Core teams, Framework, Data, CLI
+and Learning, and is usually done over the period of a week. This is so because
+some packages or projects require overs to have been published beforehand. The
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Not sure what do you mean by _overs_ here. Do you refer to dependencies/dependents?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+steps are as follows:
+
+- The Framework team releases the new version of the Ember npm package,
+  [ember-source](https://www.npmjs.com/package/ember-source).
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: 💡 I would explain what does this package contain. From the ones outside the Ember world is not obvious (I had to enter into Github's link to see Ember is a multi-package project and it's not clear to me in 2 clicks what is inside)
 
```suggestion
- The Framework team releases the new version of the Ember npm package,
  [ember-source](https://www.npmjs.com/package/ember-source) which contains .........
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+- The Data team updates their dependency on `ember-source`, tests that
+  everything is working as expected and releases a new version of
+  [ember-data](https://www.npmjs.com/package/ember-data).
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
- The Data team updates their dependency on `ember-source`, tests that
  everything is working as expected and releases a new version of
  [ember-data](https://www.npmjs.com/package/ember-data). `ember-data` is the package in charge of...
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+- The CLI team updates the `ember-source` and `ember-data` dependencies in the
+  blueprints, tests that everything works as expected and releases a new version
+  of [ember-cli](https://www.npmjs.com/package/ember-cli).
+- The Learning team releases the relevant versions of the
+  [Guides](https://guides.emberjs.com/) and the
+  [API documentation](https://api.emberjs.com/ember/release), updates the
+  [Releases page](https://emberjs.com/releases/), releases any relevant
+  [Deprecations](https://deprecations.emberjs.com/), and finally coordinates and
+  releases the [release blog post](https://blog.emberjs.com/tag/releases).
+
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Not sure if there's time but I would see valuable to illustrate a graph of dependencies between each project:

(ember-source) ---> (ember-data) --> (ember-cli)  
      |_________________________________⤴   
        
=> (Guides + API docs + Releases page + Deprecations + Blog Post)

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  everything is working as expected and releases a new version of
+  [ember-data](https://www.npmjs.com/package/ember-data).
+- The CLI team updates the `ember-source` and `ember-data` dependencies in the
+  blueprints, tests that everything works as expected and releases a new version
+  of [ember-cli](https://www.npmjs.com/package/ember-cli).
+- The Learning team releases the relevant versions of the
+  [Guides](https://guides.emberjs.com/) and the
+  [API documentation](https://api.emberjs.com/ember/release), updates the
+  [Releases page](https://emberjs.com/releases/), releases any relevant
+  [Deprecations](https://deprecations.emberjs.com/), and finally coordinates and
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I assume these projects do not have previous packages as dependencies, right?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+steps are as follows:
+
+- The Framework team releases the new version of the Ember npm package,
+  [ember-source](https://www.npmjs.com/package/ember-source).
+- The Data team updates their dependency on `ember-source`, tests that
+  everything is working as expected and releases a new version of
+  [ember-data](https://www.npmjs.com/package/ember-data).
+- The CLI team updates the `ember-source` and `ember-data` dependencies in the
+  blueprints, tests that everything works as expected and releases a new version
+  of [ember-cli](https://www.npmjs.com/package/ember-cli).
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
- The CLI team updates the `ember-source` and `ember-data` dependencies in the
  blueprints, tests that everything works as expected and if it does, releases a new version
  of [ember-cli](https://www.npmjs.com/package/ember-cli).
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+some packages or projects require overs to have been published beforehand. The
+steps are as follows:
+
+- The Framework team releases the new version of the Ember npm package,
+  [ember-source](https://www.npmjs.com/package/ember-source).
+- The Data team updates their dependency on `ember-source`, tests that
+  everything is working as expected and releases a new version of
+  [ember-data](https://www.npmjs.com/package/ember-data).
+- The CLI team updates the `ember-source` and `ember-data` dependencies in the
+  blueprints, tests that everything works as expected and releases a new version
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I assume you refer to this: https://cli.emberjs.com/release/advanced-use/blueprints/ ?
```suggestion
  [blueprints](https://cli.emberjs.com/release/advanced-use/blueprints/), tests that everything works as expected and releases a new version
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+<!--break-->
+
+![Effective infrastructure for efficient development workflows illustration](/assets/images/posts/2021-07-13-effective-infrastructure-for-efficient-development-workflows/illustration.svg#full)
+
+## How do Ember project releases work
+
+An Ember project release involves 4 different Core teams, Framework, Data, CLI
+and Learning, and is usually done over the period of a week. This is so because
+some packages or projects require overs to have been published beforehand. The
+steps are as follows:
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: 💅🏽  I would use a numbered list since each step depends on the previous one:

1. ...
2. ...
3. ...
4. ...

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+- The Learning team releases the relevant versions of the
+  [Guides](https://guides.emberjs.com/) and the
+  [API documentation](https://api.emberjs.com/ember/release), updates the
+  [Releases page](https://emberjs.com/releases/), releases any relevant
+  [Deprecations](https://deprecations.emberjs.com/), and finally coordinates and
+  releases the [release blog post](https://blog.emberjs.com/tag/releases).
+
+The release blog post marks the official release of a new project version, since
+it means that all of the sub-projects and relevant documentation is ready and
+tests for developers to use and consult.
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I don't understand what you mean by with `tests` in L45, but probably is something I'm not catching with my English... I rephrase it to show you what I'm understanding you pretended to write but I'm not sure.

```suggestion
The release blog post marks the milestones of the official release of a new project version. It means that all of the sub-projects are tested, released, they have relevant documentation so they are ready for developers to use and consult.
```

Comment (oscard0m) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Maybe it could be interesting to link an example of a blog post on Ember so people can read it and see?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+### The Learning team process
+
+As you might have noticed in the list above, the Learning team touches on
+several projects during the course of the release process. The team maintains a
+handbook with all the steps of the
+[Ember release process](https://github.com/ember-learn/handbook/blob/main/ember-releases.md).
+At the time of writing, they are:
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Added permalink of the moment you created the post, not sure if you see it valuable or is already clear with the text itself:

```suggestion
[At the time of writing](https://github.com/ember-learn/handbook/blob/3ff7f36e6848ac9118a65d86f959dca828513bdc/ember-releases.md), they are:
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+5. Deprecations
+6. Upgrade Guide
+7. Glitch Ember starter
+8. Ember Wikipedia
+9. Release bot
+
+The steps are a mix of easily automated tasks and tasks that necessitate manual
+intervention.
+
+We started automating this process by creating bash scripts for the Guides
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: We vs They depending on previous comment about the person used to write articles in simplabs' blog

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+you try to introduce error handling and better ergonomics for the caller of the
+script, it becomes considerably harder to write the necessary bash code.
+
+The other reason is that bash is not natively supported on Windows. There are ways to run bash on Windows systems, but we wanted to reduce incidental dependencies in order to make releases portable and easier for the user.
+
+So to run all of the steps you would need to clone the relevant repositories and follow the release handbook line by line to make sure you get the order correct.
+
+## Creating tool-new-release
+
+There was still a lot of automation left on the table, so I got to work figuring out how to improve the situation. When picking a technology I had already excluded Bash for the reasons covered. I also discarded Node.js as I don't find the programming model suited for these sorts of scripting tasks, and I had bad experiences previously, particularly around packages that require native extensions.
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
There was still a lot of automation left on the table, so I/we/they/impersonal? got to work figuring out how to improve the situation. When picking a technology I/we/they? had already excluded Bash for the reasons covered. I/we/they/impersonal? also discarded Node.js as I/we/they/impersonal? don't find the programming model suited for these sorts of scripting tasks, and I/we/they/impersonal? had bad experiences previously, particularly around packages that require native extensions.
```

I will stop mentioning this in the document, but I would review in general the pronoun/person to use along the document

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+
+At the time I was dabbling in Rust and had already make a couple of tools for myself using that language, so I decided to give Rust a go. Picking Rust also had the benefit of giving me a single binary that a maintainer can download and run, not having to deal with runtime and dependencies versions. And so, work started on [tool-new-release](https://github.com/ember-learn/tool-new-release/).
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
At the time I was dabbling in [Rust](https://www.rust-lang.org/) and had already make a couple of tools for myself using that language, so I decided to give Rust a go. Picking Rust also had the benefit of giving me a single binary that a maintainer can download and run, not having to deal with runtime and dependencies versions. And so, work started on [tool-new-release](https://github.com/ember-learn/tool-new-release/).
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+
+### First steps
+
+Determined to make the tool useful with as little effort as possible and then iterate, I started out by shelling out using [`std::process::Command`](https://doc.rust-lang.org/std/process/struct.Command.html) from Rust's standard library. While shelling out did not improve on the portability problem, making the tool do something useful was a motivator to keep developing it.
+
+Next I added [`structopt`](https://docs.rs/structopt/0.3.23/structopt/) to provide command-line argument parsing. This would be useful to allow the user to select which project they want to run the release steps for, to provide the target version, and more. `structopt` makes it really straightforward to implement interfaces using native Rust structures. Here's an example of a `--project` flag that accepts either `Guides` or `Api` as values:
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: 💡 I would see valuable to attach a video/screenshot of the command-line prompt in action (maybe a comparison with previous bash version if there was any?) instead of ```bash``` copy paste?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+/// Ember Learning team release helper.
+#[derive(Debug, StructOpt)]
+struct Opts {
+    /// Pick which step to run the deploy pipeline for.
+    #[structopt(short, long)]
+    project: Option<Project>,
+}
+```
+
+I want to note that this also gives us documentation:
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
I want to note that this also gives us `--help` command out of the box:
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+So to run all of the steps you would need to clone the relevant repositories and follow the release handbook line by line to make sure you get the order correct.
+
+## Creating tool-new-release
+
+There was still a lot of automation left on the table, so I got to work figuring out how to improve the situation. When picking a technology I had already excluded Bash for the reasons covered. I also discarded Node.js as I don't find the programming model suited for these sorts of scripting tasks, and I had bad experiences previously, particularly around packages that require native extensions.
+
+At the time I was dabbling in Rust and had already make a couple of tools for myself using that language, so I decided to give Rust a go. Picking Rust also had the benefit of giving me a single binary that a maintainer can download and run, not having to deal with runtime and dependencies versions. And so, work started on [tool-new-release](https://github.com/ember-learn/tool-new-release/).
+
+### First steps
+
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: What do you think about adding 4th level titles? It would be too much? I see it interesting to encapsulate these 3:
#### Shelling out
#### Providing Command-line arguments
#### Operating with Git

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+Now that I had the basics of the tool working for Guides and API documentation, it was time to make it available for other people.
+
+### Releasing the tool
+
+I wanted the tool to be useable from Linux, Windows and macOS so I had to figure out a way to build the Rust project to those targets.
+Since the project is hosted on GitHub, I could use GitHub Actions to build the project for me, and then make a release.
+
+It took some researching, but thankfully I was able to pool together a couple of articles and GitHub actions to come up with a pipeline. You can see what the GitHub Action workflows currently look like in the repository's [`workflows` directory](https://github.com/ember-learn/tool-new-release/tree/main/.github/workflows) and if you have any suggestions, please let me know! Everything about the project is very much a work in progress and not optimal.
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
It took some researching, but thankfully I was able to pool together a couple of articles and [GitHub Actions](https://github.com/features/actions) to come up with a pipeline. You can see what the GitHub Action workflows currently look like in the repository's [`workflows` directory](https://github.com/ember-learn/tool-new-release/tree/main/.github/workflows) and if you have any suggestions, please let me know! Everything about the project is very much a work in progress and not optimal.
```

Comment (oscard0m) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Which article/s?

```suggestion
It took some researching, but thankfully I was able to pool together [a couple of articles](url) and GitHub actions to come up with a pipeline. You can see what the GitHub Action workflows currently look like in the repository's [`workflows` directory](https://github.com/ember-learn/tool-new-release/tree/main/.github/workflows) and if you have any suggestions, please let me know! Everything about the project is very much a work in progress and not optimal.
```

Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
It took some researching, but thankfully I was able to pool together a couple of articles and GitHub actions to come up with a pipeline. You can see what the GitHub Action workflows currently look like in the repository's [`workflows` directory](https://github.com/ember-learn/tool-new-release/tree/main/.github/workflows) and if you have any suggestions, please [let me know](https://twitter.com/@locks)! Everything about the project is very much a work in progress and not optimal.
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+
+Two things I'd like to point out in the workflow is that I ended up using [`musl`](https://musl.libc.org/) through `rustup` in order to build the Linux binary, and that I'm currently publishing the binary as a _draft_, so it won show up on the homepage of the repository.
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
Two things I'd like to point out in the workflow are:

- I ended up using [`musl`](https://musl.libc.org/) through `rustup` in order to build the Linux binary
- I'm currently publishing the binary as a _draft_, so it won't show up on the homepage of the repository.
```

Comment (oscard0m) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Could you explain a little bit more about `musl`? Is not clear to me why you had to use it. What was the problem you were facing?

Not sure if it would be too deep for your article, but at least for my knowledge :)

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+- windows, linux, and macOS
+   -https://jondot.medium.com/building-rust-on-multiple-platforms-using-github-6f3e6f8b8458
+
+
+
+## Future work
+
+- codify Guides steps in Rust instead
+- resume feature like `git rebase`
+- more error handling
```
Comment (oscard0m) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I assume these are the WIP sections

**Review by oscard0m — COMMENTED** [review]
Nice article @locks !

It is very clear and a revelation for me specially that you did me the walkthrough project. I have enjoyed the reading a lot.

Besides the amount of comments try to not get stressed, they are minor suggestions which you can skip.

I would suggest to share it to some native-English speaker because I'm not the best one in English language as you already know xD

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 2 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: maybe "How the Ember project does releases"?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 21 above)
Comment (oscard0m) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: 💡 Not sure if some early conclusions would have sense on these first steps towards this tool or it's not the purpose. Something like: `we improved our developer experience, we improved performance, we are doing the project/process more X...`

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 2 above)
Comment (oscard0m) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I like it

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 3 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I think that's a typo. I'll tweak the sentence.

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 1 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Good point. I'll confirm and do a full on the article afterwards.

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 18 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I think it might be too busy since each section isn't that extensive?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 12 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I think it might be confusing to have two links to the handbook one after the other. I also want people to go to the most up to date release process, so I think I'll take the risk that the text is clear enough 😁

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 9 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I added direct links to the relevant app and addon blueprints. I think that's more relevant than the documentation. What do you think?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 6 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: If someone can help me put one together quickly that sounds like a good addition!

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 11 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I think I meant "ready and tested". Does that read clearly?

Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I added a link to the releases tag in the blog.

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 21 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I cleaned this up and added a "see you in the next post" closing paragraph. What do you think?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 20 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: It might be a bit too deep or too not relevant for the content. I'm also not confident in my ability to explain it 😂 I didn't take good notes at the time, but I think I needed a C standard library that worked on more platforms than whatever the default is (GNU?).

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 7 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Sorry, I'm not clear on what you mean!

Comment (oscard0m) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: What I mean is: are _Guides_, _API Documentation_, _Releases Page_ and  _Deprecations_ JavaScript projects with npm/yarn/pnpm? If so, do they have ember dependencies we previously updated in previous steps?

If yes, I was suggesting if we should mention them. If not, we can resolve this conversation.

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 9 above)
Comment (oscard0m) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Sounds perfect to me

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 21 above)
Comment (oscard0m) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Sounds good to me

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 22]
```diff
@@ -0,0 +1,239 @@
+---
+title: 'Automating Ember Core Learning Team releases with Rust'
```
Comment (marcoow) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
title: 'Automating Ember releases with Rust'
```

This would seem clearer to me – _"Ember Core Learning Team releases"_ sounds like something other that Ember is released ;)

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+bio: 'Senior Frontend Engineer, Ember Framework and Learning Core teams member'
+description:
+  'Ricardo Mendes discusses a command like tool built in Rust that automates the
+  releases process for the Ember Core Learning Team'
+og:
+  image: /assets/images/posts/2021-07-13-effective-infrastructure-for-efficient-development-workflows/og-image.png
+---
+
+It's not trivial to release a new version of Ember. This process involves
+several different core teams, Framework, Data, CLI and Learning, who are
```
Comment (marcoow) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
several different core teams – Framework, Data, CLI and Learning – who are
```

?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+responsible for several packages and resources between them. In this post
+we will go through the process of building a tool in Rust to automate the
+process for the Core Learning Team.
```
Comment (marcoow) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I think it'd be good to make clear in this sentence that you have in fact already done this and the Rust tool you're describing is what's being used right now to release versions of Ember – I think that's an interesting fact that will excite readers.

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+## Creating tool-new-release
+
+There was still a lot of automation left on the table, so I got to work figuring
+out how to improve the situation. When picking a technology I had already
+excluded Bash for the reasons covered. I also discarded Node.js as I don't find
+the programming model suited for these sorts of scripting tasks, and I had bad
+experiences previously, particularly around packages that require native
+extensions.
+
+At the time I was dabbling in Rust and had already make a couple of tools for
```
Comment (marcoow) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
At the time I was dabbling in Rust and had already made a couple of tools for
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+folder.push("guides-source");
+let repo = Repository::clone("https://github.com/ember-learn/guides-source.git", &folder)?;
+```
+
+In the actual tool a temporary folder is created at the top that is then fed to
+the different steps so they can manage the cloning they need to do.
+
+#### Shelling out
+
+Determined to have a viable release tool as quickly as possible as possible and
```
Comment (marcoow) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]:
```suggestion
Determined to have a viable release tool as quickly as possible and
```

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 6 above)
Comment (marcoow) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: @locks if you make a sketch on paper and take a photo, you can send that to @emmasofiah2o and she can have the designer make a proper graphic out of it.

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 27]
```diff
@@ -0,0 +1,239 @@
...
+author: 'Ricardo Mendes'
+github: locks
+twitter: locks
+topic: ember
+bio: 'Senior Frontend Engineer, Ember Framework and Learning Core teams member'
+description:
+  'Ricardo Mendes discusses a command like tool built in Rust that automates the
+  releases process for the Ember Core Learning Team'
+og:
+  image: /assets/images/posts/2021-07-13-effective-infrastructure-for-efficient-development-workflows/og-image.png
```
Comment (marcoow) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Let's make a new illustration for this. @emmasofiah2o can you have the Upwork designer make one? @locks can give you some info on what would be suitable.

**Review by marcoow — COMMENTED** [review]
Nice post! I like it – just left a few comments on some points.

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 22 above)
Comment (Turbo87) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I think that's the case though?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 28]
```diff
@@ -0,0 +1,239 @@
...
+process for the Core Learning Team.
+
+<!--break-->
+
+![Effective infrastructure for efficient development workflows illustration](/assets/images/posts/2021-07-13-effective-infrastructure-for-efficient-development-workflows/illustration.svg#full)
+
+## Release process for the Ember project
+
+An Ember project release involves 4 different Core teams, Framework, Data, CLI
+and Learning. The release is usually done over the period of a week, because
```
Comment (Turbo87) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: listing the teams here again feels a bit repetitive 😅

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 29]
```diff
@@ -0,0 +1,239 @@
...
+follow the release handbook line by line to make sure you get the order correct.
+
+## Creating tool-new-release
+
+There was still a lot of automation left on the table, so I got to work figuring
+out how to improve the situation. When picking a technology I had already
+excluded Bash for the reasons covered. I also discarded Node.js as I don't find
+the programming model suited for these sorts of scripting tasks, and I had bad
+experiences previously, particularly around packages that require native
+extensions.
```
Comment (Turbo87) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: > as I don't find the programming model suited for these sorts of scripting tasks

can you elaborate on that? on first glance a scripting language like JS would seem more appropriate for "scripting tasks", especially since Ember itself is also JS and you wouldn't have to train devs on how to work on the tool

**Review by Turbo87 — APPROVED** [review]

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 29 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Hm, I didn't ellaborate kind of on purpose to stay on topic, so I should tweak this a little bit. Basically whenever I try to do anything in Node it's a pain in the ass having to deal with callbacks/async calls and dealing with processes.

Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Maybe I should focus more on it being more work bundling Node scripts in an executable so people don't have to install Node or deal with versions?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 22 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I'm conflicting on this. The tool automates parts of the Learning team process, so in a sense it's not releasing Ember itself. It's the Guides, the API Documentation, the Releases page, etc.

Comment (marcoow) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I think many readers even within the Ember community will not know what the Core Learning team even is though

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 7 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I see. It's not an npm dependency, no, just a procedural one.

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md (hunk 28 above)
Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: I removed the team listing from the intro so I can keep it here. Does that address your concern?

File: _posts/2021-10-18-automating-ember-learning-releases-with-rust.md [hunk 30]
```diff
@@ -0,0 +1,237 @@
...
+  releases process for the Ember Core Learning Team'
+og:
+  image: /assets/images/posts/2021-10-18-automating-ember-learning-releases-with-rust/og-image.png
+---
+
+It's not trivial to release a new version of Ember. This process involves
+several different core teams, who are responsible for several packages and
+resources between them. In this post we will go through how I used Rust to build
+the tool that is used to automate part of the Ember release.
+
```
Comment (emmasofiah2o) [inline _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: How I'd put the intro: "Releasing a new version of Ember is no uncomplicated task. This process involves several different core teams, who are responsible for several packages and resources between them. In this post, I will explain how I used Rust to build a tool that automates part of the Ember release process."

Comment (locks) [reply _posts/2021-10-18-automating-ember-learning-releases-with-rust.md]: Updated!
